docs_src/design/adr/013-Device-Service-Events-Message-Bus.md

```diff
@@ -0,0 +1,252 @@
+
+> *Note: The current `FilterByDeviceName` is only good if the device name is known statically and the only instance of the device defined by the `DeviceProfileName`. What we really need is `FilterByDeviceProfileName` which allows multiple instances of a device to be filtered for, rather than a single instance as it it now. The V2 API will be adding `DeviceProfileName` to the Events, so in Ireland this  filter will be possible.*
+
+Pub/Sub systems have advanced topic schema, which we can take advantage of from Application Services to filter for just the Events the Application Service actual wants. Publishers of Events must add the `DeviceProfileName` and/or `DeviceName` to the topic in the form `edgex/events/<device-profile-name>/<device-name>`. This allows Application Services to filter for just the Events from the device(s) it wants by only subscribing to those `DeviceProfileNames` or the specific `DeviceNames`. Example subscribe topics if above schema is used:
+
+- **edgex/events/#**
+  - All Events 
+  - Core Data will subscribe using this topic schema
+- **edgex/events/Random-Integer-Device/#** 
+  - Only Events from devices created from the **Random-Integer-Device** device profile
+- **edgex/events/Random-Integer-Device/Random-Integer-Device1** 
+  - Only Events from the **Random-Integer-Device1** Device
+
+The MessageBus abstraction allows for multiple subscriptions, so an Application Service could specify to receive data from multiple specific device profiles or devices by creating multiple subscriptions. i.e.  `edgex/Events/Random-Integer-Device/#` and  `edgex/Events/Random-Boolean-Device/#`. Currently the App SDK only allows for a single subscription topic to be configured, but that could easily be expanded to handle a list of subscriptions. See [Configuration](#configuration) section below for details. 
+
+Core Data's existing publishing of Events would also need to be changed to use this new topic schema. One challenge with this is Core Data doesn't currently know the `DeviceProfileName` or `DeviceName` when it receives a CBOR encoded event. This is because it doesn't decode the Event until after it has published it to the MessageBus. The V2 API will be enhanced to change the AddEvent endpoint from `/event` to `/event/{profile}/{device}` so that `DeviceProfileName` and `DeviceName` are always know no matter how the request is encoded.
+
+This new topic approach will be enabled via each publisher's `PublishTopic` having the placeholder(s) for the `DeviceProfileName` and `DeviceName` which get replaced with the actual  `DeviceProfileName` and `DeviceName`.
+
+```toml
+PublishTopic = 'edgex/events/<device-profile-name>'
+or
+PublishTopic = 'edgex/events/<device-profile-name>/<device-name>'
+```
+
```

Comment:
I'm concerned that having the device and profile names in the topic is going to be very useful and we ought to mandate it. So maybe allow the topic prefix to be changed from `edgex/events` but always have the names on the end

Reply:
@iain-anderson , So `PublishTopic `, be comes `PublishTopicPrefix` and doesn't contain the place holders since they will always be there?
Something like this?
```
PublishTopicPrefix = 'edgex/events`  # /<device-profile-name>/<device-name> will be added to this prefix
```
I am good with that.

Reply:
Core WG agreed with this approach. ADR has been updated.